feat(view): add a pickFeaturesAt method in View

zarov opened the pull request:
This new method comes in addition to `pickObjectsAt`: it looks for all
features under the given screen coordinate, and return it sorted by
layer.

In the same time, `FeatureToolTip` has been greatly improved.

zarov committed:
feat(view): add isGlobeView and isPlanarView booleans

gchoqueux commented on examples/js/plugins/FeatureToolTip.js:
```diff
@@ -1,150 +1,224 @@
+
+            layer = layers[layersId.indexOf(layerId)];
+            if (typeof layer.options.filterGeometries == 'function') {
+                features[layerId] = layer.options.filterGeometries(features[layerId], layer.layer);
```
I prefer that the user gives a function `getIdFromProperties(properties)` and that we develop with a filtering function.
May be throw a error if this function isn't exist.

zarov replied:
The idea here is that the user could filter the features as he wishes, not necessary by using the id.

gchoqueux replied:
ok

gchoqueux commented on examples/js/plugins/FeatureToolTip.js:
```diff
@@ -1,150 +1,224 @@
+                content += '<br/><span class="coord">lat ' + feature.coordinates[1].toFixed(4) + '</span>';
+            }
+
+            if (geometry.properties && !options.filterAllProperties) {
```
`filterAllProperties` could explain this option?

zarov replied:
This is the previous `filterAll` option (see the documentation below)

gchoqueux replied:
yes, the documentation doesn't give a real explanation.

zarov replied:
Yes, updated

gchoqueux commented on src/Core/View.js:
```diff
@@ -636,6 +637,75 @@ class View extends THREE.EventDispatcher {
+        };
+
+        // Get the mouse coordinates to the correct system
+        const positionVector = new THREE.Vector3();
```
could you instance the `positionVector` and `position` before function?

zarov replied:
Yes !

gchoqueux commented on src/Core/View.js:
```diff
@@ -636,6 +637,75 @@ class View extends THREE.EventDispatcher {
+        let precision;
+        const precisions = {
+            PX: radius,
+            M: this.isGlobeView && this.controls ? this.controls.pixelsToMeters(radius) : radius,
```
the precision in meter in wrong for planar view.

zarov replied:
Yes, I forgot to look at this

gchoqueux commented on src/Core/View.js:
```diff
@@ -636,6 +637,75 @@ class View extends THREE.EventDispatcher {
+        const tiles = Picking.pickTilesAt(this, mouse, precisions.PX, this.tileLayer);
+
+        for (const t in tiles) {
+            if (!tiles[t].object) {
```
`!tiles[t].object` What is this case?

zarov replied:
None, I'm removing it, it something I forgot

gchoqueux commented on src/Core/View.js:
```diff
@@ -636,6 +637,75 @@ class View extends THREE.EventDispatcher {
+                continue;
+            }
+
+            for (const materialLayer of tiles[t].object.material.layers) {
```
could use loop on `layers` and use `material.getLayer(id)` to get `parsedData`

zarov replied:
Done

gchoqueux replied:
`for (const material of tile.object.material.getLayers(layers)) {`

just rename  `material` to `materialLayer`

zarov replied:
Done

gchoqueux commented on src/Core/View.js:
```diff
@@ -636,6 +637,75 @@ class View extends THREE.EventDispatcher {
+
+                    precision = CRS.isMetricUnit(texture.parsedData.crs) ? precisions.M : precisions.D;
+                    FeaturesUtils.filterFeaturesUnderCoordinate(position, texture.parsedData, precision)
+                        .forEach(f => result[materialLayer.id].push(f));
```
could use `Array.concat`?

zarov replied:
Ok

gchoqueux reviewed (commented):
Thanks, the performance is good.
After this PR, could you do a PR to apply this feature on vector tile?

zarov commented:
Updated

zarov commented:
I rebased, and added a commit that refactor the whole meter/degree/pixel system to a more simple and generic one. I also marked some methods in `GlobeControls` as deprecated, as I think we can remove them in the next release (or the one after).

I will add some functional testing to really validate those conversions, but for now it seems to give correct values.

zarov commented:
Bad news, the `getMetersToPixels` and `getPixelsToMeters` don't return the same value, I'll get back to it.

gchoqueux commented on src/Utils/FeaturesUtils.js:
```diff
@@ -144,8 +145,23 @@ export default {
+            // Special case, because of the way tiles in VectorTileParser are
+            // handled (see Feature2Texture for a similar solution)
+            if (features.extent.crs == 'WMTS:PM') {
+                const ex = new Extent(coordinate.crs, 0, 0, 0, 0);
```
could you instance ex before function?

zarov replied:
done

gchoqueux commented on src/Utils/FeaturesUtils.js:
```diff
@@ -144,8 +145,23 @@ export default {
+        if (features.extent) {
+            // Special case, because of the way tiles in VectorTileParser are
+            // handled (see Feature2Texture for a similar solution)
+            if (features.extent.crs == 'WMTS:PM') {
```
it isn't the PM case, it's the case when features have scale or/and translation transformation.

zarov replied:
done

gchoqueux commented on src/Core/View.js:
```diff
@@ -636,6 +654,120 @@ class View extends THREE.EventDispatcher {
+    getPixelsToMeters(pixels = 1, screenCoord) {
+        const br = this.mainLoop.gfxEngine.renderer.domElement.getBoundingClientRect();
+
+        let _screenCoord = (screenCoord instanceof Event) ? this.eventToViewCoords(screenCoord) : screenCoord;
```
Could you use `const meter = pixel * D / this.camera._preSSE;` instead of?

zarov replied:
Done

gchoqueux commented on src/Core/Prefab/GlobeView.js:
```diff
@@ -151,6 +152,14 @@ class GlobeView extends View {
         return super.addLayer(layer, this.tileLayer);
     }
+
+    getPixelsToDegrees(pixels, screenCoord) {
```
There's no need a specific function for `GlobeView`

zarov replied:
Why ?

gchoqueux reviewed (changes requested):
It's work perfectly.

zarov commented:
Rebased and updated !

gchoqueux commented on src/Controls/GlobeControls.js:
```diff
@@ -1105,6 +1105,9 @@ GlobeControls.prototype.setZoom = function setZoom(zoom, isAnimated) {
  * @param      {number}  pitch   Screen pitch, in millimeters ; 0.28 by default
  * @return     {number}  The zoom scale.
+ *
+ * @deprecated To remove with pixelsToMeters, pixelsToDegrees and
```
`getScale` will have to be moved to view

zarov replied:
done

gchoqueux replied:
use `view.getScale()` inside `GlobeControls.getScale()`

gchoqueux commented on src/Core/View.js:
```diff
@@ -636,6 +640,123 @@ class View extends THREE.EventDispatcher {
+        if (!_screenCoord) {
+            // TODO: This is totally arbitrary, there can be nothing here. Maybe
+            // find a better solution for this.
+            _screenCoord = new THREE.Vector2(br.width / 2, br.height / 2);
```
could you use `screen` instead of new instance.

zarov replied:
done

gchoqueux commented on src/Core/View.js:
```diff
@@ -636,6 +640,123 @@ class View extends THREE.EventDispatcher {
+        }
+
+        this.getPickingPositionFromDepth(_screenCoord, positionVector);
+        position.setFromVector3(positionVector);
```
why you copy `position` from `positionVector`?

zarov replied:
to be used in `pickFeaturesAt` but I modified that

gchoqueux commented on src/Utils/FeaturesUtils.js:
```diff
@@ -144,8 +147,33 @@ export default {
+                }
+
+                if (!features.translation) {
+                    features.translation = new THREE.Vector3();
```
could you avoid new instance?

zarov replied:
done

gchoqueux replied:
this is useless `features.translation` ans `features.scale` have default values.

zarov committed:
feat(view): add a pickFeaturesAt method in View

This new method comes in addition to pickObjectsAt: it looks for all
features under the given screen coordinate, and return it sorted by
layer.

This method comes with some others, that were present in GlobeControls
only and are now generalised: View#getPixelsToMeters,
View#getMetersToPixels, View#getScale and View#getDistanceFromCamera.

zarov committed:
feat(plugins): readapt FeatureToolTip with pickFeaturesAt

Since the last commit introducing pickFeaturesAt, this plugin has been
rewritten and documentation has been added.

zarov commented:
Rebased and cleaned !
